File: semana 6/adapter.py
class PasarelaUSD:
    def procesar_pago(self, monto_usd: float) -> str:
        return f"Cobrado {monto_usd:.2f} USD"


class AdaptadorSoles:
    """Adapter que convierte Soles a USD"""

    def __init__(self, pasarela: PasarelaUSD, tasa: float = 3.70):
        self.pasarela = pasarela   # Composición
        self.tasa = tasa

    def procesar_pago_soles(self, monto_pen: float) -> str:
        # Conversión: interfaz incompatible -> compatible
        usd = monto_pen / self.tasa
        return self.pasarela.procesar_pago(usd)


if __name__ == "__main__":
    print("=== Adapter Pattern ===")

    api_usd = PasarelaUSD()
    adapter = AdaptadorSoles(api_usd, 3.70)

    print(f"Pago en soles (100 PEN): {adapter.procesar_pago_soles(100.0)}")

    print("\n[Nota] Python permite duck typing:")
    print("No necesitas heredar de interfaz específica.")

File: semana 6/adapter.cpp
#include <iostream>
#include <string>
using namespace std;

// Interfaz existente incompatible
class PasarelaUSD {
public:
    string procesar_pago(double monto_usd) {
        return "Cobrado " + to_string(monto_usd) + " USD";
    }
};

// Interfaz que el cliente necesita
class PagosSoles {
public:
    virtual string procesar_pago_soles(double monto_pen) = 0;
    virtual ~PagosSoles() = default;
};

// Adapter: convierte Soles a USD
class AdaptadorSoles : public PagosSoles {
private:
    PasarelaUSD* pasarela;
    double tasa; // Tipo de cambio

public:
    AdaptadorSoles(PasarelaUSD* p, double t = 3.70)
        : pasarela(p), tasa(t) {}

    string procesar_pago_soles(double monto_pen) override {
        double usd = monto_pen / tasa;
        return pasarela->procesar_pago(usd);
    }
};

int main() {
    cout << "=== Adapter Pattern ===" << endl;

    PasarelaUSD api_usd;
    AdaptadorSoles adapter(&api_usd, 3.70);

    cout << "Pago en soles (100 PEN): "
         << adapter.procesar_pago_soles(100.0) << endl;

    cout << "\n[Nota] El adapter permite usar PasarelaUSD "
         << "sin modificar su código, solo adaptando la interfaz." << endl;

    return 0;
}
